### PR TITLE
feat: use proxy instead of port forwards

### PR DESCRIPTION
**Pull request by smuu**
## Overview

This pull request removes the need for port-forwarding. Instead, it facilitates the reverse proxy that is available in the scope.

I will comment in code to explain the changes.

@staheri14 Feel free to merge this as well to your branch

### CODE REVIEW

File: test/e2e/check_upgrades.go
Comment (smuu) [inline test/e2e/check_upgrades.go:233]: Changes it back to how it was done before.

File: test/e2e/simple.go
Comment (smuu) [inline test/e2e/simple.go:56]: I don't understand this. It failed me, as the error returned is not `context.DeadlineExceeded`. Commented it out for now.

File: test/e2e/testnet/node.go
Comment (smuu) [inline test/e2e/testnet/node.go:280]: This returns a host to reach the RPC/gRPC endpoint via the reverse proxy without relying on port-forwards.

File: test/e2e/testnet/node.go
Comment (smuu) [inline test/e2e/testnet/node.go:298]: We don't need to forward ports anymore.

File: test/e2e/testnet/node.go
Comment (smuu) [inline test/e2e/testnet/node.go:315]: This is not needed. When using `instance.EnableBittwister` this is done under the hood.

File: test/e2e/testnet/node.go
Comment (smuu) [inline test/e2e/testnet/node.go:330]: Not needed anymore.

File: test/e2e/testnet/testnet.go
Comment (smuu) [inline test/e2e/testnet/testnet.go:23]: The Executor is not needed anymore.

File: test/e2e/testnet/testnet.go
Comment (smuu) [inline test/e2e/testnet/testnet.go:333]: There are scenarios where the reverse proxy still needs to be ready. This does not fail now when the status requests return a non `2xx` return code.

File: test/e2e/testnet/node_helpers.go
Comment (smuu) [inline test/e2e/testnet/node_helpers.go:1]: This was a workaround and is not needed anymore.

File: go.mod
Comment (ninabarbakadze) [inline go.mod:11]: thanks for the pr @smuu 🙏🏻
when are you guys planning on releasing this version? I'd prefer that over using an rc.

File: test/e2e/check_upgrades.go
Comment (rach-id) [inline test/e2e/check_upgrades.go:238]: I guess this should also check: `case <-ctx.Done()` in case some cancel happened for the parent context somewhere

File: test/e2e/simple.go
Comment (rach-id) [inline test/e2e/simple.go:59]: why are we keeping commented code?

File: test/e2e/testnet/node.go
Comment (rach-id) [inline test/e2e/testnet/node.go:223]:
```suggestion
// This returns the local proxy host that can be used to communicate with the node
```

File: test/e2e/testnet/node.go
Comment (rach-id) [inline test/e2e/testnet/node.go:45]: The ports are no longer needed?

File: test/e2e/testnet/node.go
Comment (rach-id) [inline test/e2e/testnet/node.go:229]:
```suggestion
// local proxy host that can be used to communicate with the node
```

File: test/e2e/testnet/testnet.go
Comment (rach-id) [inline test/e2e/testnet/testnet.go:332]: [nit][optional]
is there a chance the context gets cancelled somewhere else? If so, it would be good to have a ticker here along with a `case <-ctx.Done()` in case the context was done somewhere else and this stays pending for 9sec before it stops.

**Review by rach-id — COMMENTED** [review]
Overall LGTM 🎉 

Left some comments to understand more.

**Review by evan-forbes — APPROVED** [review]
lgtm! second some of the nits from Rachid and Nina

File: test/e2e/simple.go
Comment (rootulp) [reply test/e2e/simple.go:56]: +1 on not understanding this. I created https://github.com/celestiaorg/celestia-app/issues/3380 which we can close if we end up deleting these lines.

File: go.mod
Comment (smuu) [reply go.mod:11]: The latest beginning of next week.

File: test/e2e/testnet/node.go
Comment (smuu) [reply test/e2e/testnet/node.go:45]: As we use the proxy host, the proxy ports do not exist anymore.

File: test/e2e/testnet/testnet.go
Comment (smuu) [reply test/e2e/testnet/testnet.go:332]: The context is created in the scope of the function with `context.Background()`.

**Review by rach-id — DISMISSED** [review]

File: go.mod
Comment (smuu) [reply go.mod]: It uses the release `v0.14.0` now.

**Comment by smuu** [comment]
This PR now uses the release `v0.14.0`, and all conflicts are resolved.
Feel free to merge this PR.

File: test/e2e/simple.go
Comment (cmwaters) [reply test/e2e/simple.go:56]: What error did you get? Was it context cancelled?

Comment (cmwaters) [reply test/e2e/simple.go:56]: This should always return an error

File: test/e2e/testnet/testnet.go
Comment (cmwaters) [reply test/e2e/testnet/testnet.go:334]: Is there a way where the reverse proxy can tell us when it's ready? Or a way where we don't need to rely on sleeping

**Review by cmwaters — DISMISSED** [review]
Thanks @smuu! I have a few comments but these can also be addressed in a follow up PR as to not be blocking

File: test/e2e/testnet/testnet.go
Comment (smuu) [reply test/e2e/testnet/testnet.go:334]: @mojtaba-esk

Comment (mojtaba-esk) [reply test/e2e/testnet/testnet.go:334]: When `Knuu` is initialized with proxy enabled, the deploy and check service for proxy is done internally and they are blocking, so once knuu is initialized, the proxy should be ready. There should be no need for sleep and wait.

here: https://github.com/celestiaorg/knuu/blob/4e6dcf097463fd9a58c40d2d7c7c9065ac94d9cc/pkg/traefik/traefik.go#L177-L178

Comment (smuu) [reply test/e2e/testnet/testnet.go:334]: The deployment of the proxy is blocking until deplopyed.
But when doing some testing I observed that the proxy takes some time to apply the config for a certain Instance.
When creating a forward for an Instance using the proxy, it makes a custom resource in Kubernetes, which is then picked up by the proxy.
Sometimes, we use the proxy host before the proxy applies this configuration.

Comment (mojtaba-esk) [reply test/e2e/testnet/testnet.go:334]: oh I see, so we probably can add something to the `AddHost` func to check that. or have a callback func if that's needed.

Comment (mojtaba-esk) [reply test/e2e/testnet/testnet.go:334]: would you mind to create an issue for it to keep tracking of it in its own issue?

File: test/e2e/simple.go
Comment (cmwaters) [reply test/e2e/simple.go:56]: cc @smuu (just so we don't forget this)

File: test/e2e/testnet/testnet.go
Comment (mojtaba-esk) [reply test/e2e/testnet/testnet.go:334]: @smuu

**Review by staheri14 — CHANGES_REQUESTED** [review]
I attempted to test the e2e tests using the current branch, but encountered an error about the nmt version specified in the go.mod which is not valid (as far as I checked nmt releases, this version `0.20.1` does not exist), so, needed to request for changes to fix it before merge. Other than that, it looks good to me. Thanks!

File: test/e2e/testnet/testnet.go
Comment (smuu) [reply test/e2e/testnet/testnet.go:334]: Created issue in knuu: https://github.com/celestiaorg/knuu/issues/423

File: test/e2e/simple.go
Comment (smuu) [reply test/e2e/simple.go:56]: I don't get any error when running txsim.
When having the code to expect `context.DeadlineExceeded`, the test fails with this specified error.

Comment (smuu) [reply test/e2e/simple.go:56]: @cmwaters What was the logic for you to add this check?

**Comment by smuu** [comment]
When running `MinorVersionCompatibility` I get the following error:

```
test-e2e2024/06/06 09:32:09 --- ERROR MinorVersionCompatibility: expected context.Canceled error, got: no suitable master account found
```

@staheri14 can you help me fix this? It's the first time I've seen this issue. and I'm not sure what broke the test.

File: go.mod
Comment (smuu) [reply go.mod]: I was calling `go mod tidy` and now it uses version `github.com/celestiaorg/nmt v0.21.0` again

**Review by staheri14 — DISMISSED** [review]

**Comment by staheri14** [comment]
> When running `MinorVersionCompatibility` I get the following error:
> 
> ```
> test-e2e2024/06/06 09:32:09 --- ERROR MinorVersionCompatibility: expected context.Canceled error, got: no suitable master account found
> ```
> 
> @staheri14 can you help me fix this? It's the first time I've seen this issue. and I'm not sure what broke the test.

As the first step, mind merging the main branch into your branch?

**Comment by cmwaters** [comment]
@smuu do you mind fixing the linting issues here and then I will give the PR another pass

**Comment by staheri14** [comment]
Some update: I have been running the `MinorVersionCompatibility` test,  and debugging. Though, there is an issue which slows down the debugging process and that is  when I attempt a second run of the test immediately after the first one finishes, I consistently get this error: 
```
failed to create testnet: cannot initialize knuu: cannot deploy Traefik: error creating Traefik deployment: deployments.apps "traefik-deployment" already exists exit status 1
```
 This error persists for some time, and then at some random point in time it does not show up. I also tried deleting replica sets and pods using the appropriate command in my knuu namespace (as the result of which I am prompted with `replicaset.apps "traefik-deployment-5b896584fd" deleted`, but even after this, the error still doesn't go away. Is this a known issue, or is there a way to resolve it?

**Comment by mojtaba-esk** [comment]
> Some update: I have been running the `MinorVersionCompatibility` test, and debugging. Though, there is an issue which slows down the debugging process and that is when I attempt a second run of the test immediately after the first one finishes, I consistently get this error:
> 
> ```
> failed to create testnet: cannot initialize knuu: cannot deploy Traefik: error creating Traefik deployment: deployments.apps "traefik-deployment" already exists exit status 1
> ```
> 
> This error persists for some time, and then at some random point in time it does not show up. I also tried deleting replica sets and pods using the appropriate command in my knuu namespace (as the result of which I am prompted with `replicaset.apps "traefik-deployment-5b896584fd" deleted`), but even after this, the error still doesn't go away. Is this a known issue, or is there a way to resolve it?

It probably resuses a namespace. try to generate a new psudo random namespace for each test run.

**Comment by staheri14** [comment]
@smuu I furthered my debugging and here are some more updates which better illustrates the cause of the test failure:
- The error is not specific to the `MinorVersionCompatibility`, it also happens for `E2ESimple`, and perhaps for other tests as well. 
- The issue is with the grpc address from which the txsim attempts to query the accounts balance i.e., this [line](https://github.com/smuu/celestiaorg-celestia-app/blob/c0904490d3e60a8206b43728378b4ae4c097edd1/test/txsim/account.go#L110). The error, given below, indicates that the address has an incorrect format, i.e., there is an excess port number `:443` at the end of it `http://151.115.13.204:80/val0-3410e0a4-9090:443`. 
Full log:
 ```
{"level":"error","error":"error getting balance for celestia1p604nkecpkuc3afm5lmrl6rrmzc3ruhr6f2nr9: failed to exit idle mode: invalid target address http://151.115.13.204:80/val0-3410e0a4-9090, error info: address http://151.115.13.204:80/val0-3410e0a4-9090:443: too many colons in address","account":"alice","time":"2024-06-11T15:56:44-07:00","message":"error getting initial account balance"}
```
To see the logs of txsim you need to remove the `SuppressLogs()`  from the txsim [options](https://github.com/celestiaorg/celestia-app/blob/c43eb0a368604600904db2dfdd005fb097c69460/test/e2e/simple.go#L51)  i.e., `opts := txsim.DefaultOptions().WithSeed(seed)`

**Comment by cmwaters** [comment]
> When running `MinorVersionCompatibility` I get the following error:
> 
> ```
> test-e2e2024/06/06 09:32:09 --- ERROR MinorVersionCompatibility: expected context.Canceled error, got: no suitable master account found
> ```
> 
> @staheri14 can you help me fix this? It's the first time I've seen this issue. and I'm not sure what broke the test.

I also get this error. I'm looking into it

**Comment by cmwaters** [comment]
It seems like the grpc library appends the port :443

**Comment by cmwaters** [comment]
@smuu is it possible that knuu doesn't add a path to the grpc url like: 

```
http://151.115.13.204:80/val0-f50d1ae0-9090
```

and has something similar to how it was originally:

```
127.0.0.1:52845
```

**Comment by smuu** [comment]
Txsim is now deployed as an instance.
As gRPC is currently not possible via the reverse proxy, I disabled it -> @evan-forbes. We need to create follow-up issues for that.

All tests are passing now on my end.

**Comment by cmwaters** [comment]
> As gRPC is currently not possible via the reverse proxy, I disabled it -> @evan-forbes. We need to create follow-up issues for that.

This just means that all assertions on the network have to be done using RPC right?

**Comment by cmwaters** [comment]
Looks like the linter is complaining about a bunch of deprecated methods in knuu. Do you mind updating them @smuu and then updating the branch? 🙏

**Comment by smuu** [comment]
> Looks like the linter is complaining about a bunch of deprecated methods in knuu. Do you mind updating them @smuu and then updating the branch? 🙏

@cmwaters This would mean a major rewrite of your usage of knuu. We paused that for now, as the repo has so many PRs open that is touching the e2e tests.
Using the new structure would blow this PR.

I propose to ignore the deprecation warning in this PR, get all the knuu-related PRs merged, and then create a PR to move to the new structure.

**Comment by cmwaters** [comment]
Okay, I will disable `staticcheck` on these files for now

File: test/e2e/simple.go
Comment (staheri14) [inline test/e2e/simple.go:34]: [Not to be addressed in this PR yet relevant]: `testnet.TxsimVersion` is currently a hardcoded version of txsim. While it works fine for benchmark tests, I believe that for  e2e tests, especially considering our plan to add [this capability](https://github.com/celestiaorg/celestia-app/issues/3509), we need to ensure that we always use the latest version of txsim. We already have an issue for this: https://github.com/celestiaorg/celestia-app/issues/3288.

**Review by staheri14 — APPROVED** [review]
Thanks @smuu! I ran all the e2e tests and confirm that they work:
- E2ESimple ✅ 
- MinorVersionCompatibility ✅ 
- MajorUpgradeToV2 ✅ 


The only concern I have is that I keep requiring creating a new namespace for each test (as suggested by @mojtaba-esk  in this [comment](https://github.com/celestiaorg/celestia-app/pull/3505#issuecomment-2160141734)) due to this error 
```
2024/06/17 12:20:09 failed to create testnet: cannot initialize knuu: cannot deploy Traefik: error creating Traefik deployment: deployments.apps "traefik-deployment" already exists
```
Could it be related to the changes introduced in this PR? Does it also happen to other reviewers?

**Review by evan-forbes — APPROVED** [review]
@smuu can we link any issues to re-enable reverse proxy (or whatever mechanism enables us to access grpc via the node assessing the tests) for posterity? 

thanks!

**Review by cmwaters — APPROVED** [review]
👍 thanks @smuu this is super helpful

**Comment by cmwaters** [comment]
@evan-forbes will the local machine running the tests ever need access to a nodes grpc? (Can't we use rpc or grpc gateway instead)

**Comment by smuu** [comment]
> Thanks @smuu! I ran all the e2e tests and confirm that they work:
> 
>     * E2ESimple ✅
> 
>     * MinorVersionCompatibility ✅
> 
>     * MajorUpgradeToV2 ✅
> 
> 
> The only concern I have is that I keep requiring creating a new namespace for each test (as suggested by @mojtaba-esk in this [comment](https://github.com/celestiaorg/celestia-app/pull/3505#issuecomment-2160141734)) due to this error
> 
> ```
> 2024/06/17 12:20:09 failed to create testnet: cannot initialize knuu: cannot deploy Traefik: error creating Traefik deployment: deployments.apps "traefik-deployment" already exists
> ```
> 
> Could it be related to the changes introduced in this PR? Does it also happen to other reviewers?

@staheri14 How are you running the tests?

It's the default now for knuu to have a namespace per test run.

**Comment by smuu** [comment]
> @smuu can we link any issues to re-enable reverse proxy (or whatever mechanism enables us to access grpc via the node assessing the tests) for posterity?
> 
> thanks!

To my understanding out reverse proxy can proxy gRPC connections.
The issue I see is the gRPC client as it appends the port `443` even though we already have a port and a path assigned. (Ref.: https://github.com/celestiaorg/celestia-app/pull/3505#issuecomment-2162664190) Your RPC client is handling it as I would expect.
So in my view, we would need to create an issue in celestia-app repo to fix that.

**Comment by smuu** [comment]
@evan-forbes @cmwaters, as I don't have permission to merge, feel free to merge this when you feel the PR is ready.

**Comment by cmwaters** [comment]
Created this issue: https://github.com/celestiaorg/celestia-app/issues/3580

**Comment by cmwaters** [comment]
@staheri14, I'm also seeing the namespace problem appear in our CI tests:

```
test-e2e2024/06/18 13:12:01 === RUN MinorVersionCompatibility
time="2024-06-18T13:12:01Z" level=info msg="Initializing knuu with scope: runMinorVersionCompatibility_20240618_131201"
time="2024-06-18T13:12:01Z" level=info msg="The .env file does not exist, continuing without loading environment variables."
test-e2e2024/06/18 13:12:01 Running minor version compatibility test versions v1.0.0	v1.1.0	v1.10.0	v1.10.1	v1.11.0	v1.2.0	v1.3.0	v1.4.0	v1.5.0	v1.6.0	v1.7.0	v1.8.0	v1.9.0	v2.0.0-rc0	v2.0.0-rc1
time="2024-06-18T13:12:01Z" level=info msg="LOG_LEVEL: info" file="knuu/knuu.go:264"
2024/06/18 13:22:01 failed to create testnet: cannot initialize knuu: cannot deploy Traefik: waiting for deployment traefik-deployment to be ready
```

Going to create an issue.

**Pull request merged**
